gate-deck-plugins/src/main/kotlin/com/netflix/spinnaker/gate/plugins/DeckPluginCache.kt

```diff
@@ -0,0 +1,146 @@
+ * TODO(rz): Downloads all plugins to an isolated temp directory, so if Gate itself has plugins, this cache will be
+ *  downloading an intersection of plugins twice.
+ */
+class DeckPluginCache(
+  private val updateManager: SpinnakerUpdateManager,
+  private val updateService: PluginUpdateService,
+  private val pluginBundleExtractor: PluginBundleExtractor,
+  private val registry: Registry
+) {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  private val cache: MutableSet<PluginCacheEntry> = mutableSetOf()
+  private var cachePopulated: Boolean = false
+
+  private val versionsId = registry.createId("plugins.deckCache.versions")
+  private val hitsId = registry.createId("plugins.deckCache.hits")
+  private val missesId = registry.createId("plugins.deckCache.misses")
+  private val downloadDurationId = registry.createId("plugins.deckCache.downloadDuration")
+  private val refreshDurationId = registry.createId("plugins.deckCache.refreshDuration")
+
+  /**
+   * Refreshes the local file cache of _current_ plugins. Should Deck need plugin assets from an older plugin release
+   * version, it will be downloaded and cached on-demand.
+   *
+   * The default refresh interval is 5 minutes.
+   */
+  @Scheduled(
+    fixedDelayString = "\${spinnaker.extensibility.deck-proxy.cache-refresh-interval-ms:300000}",
+    initialDelay = 0
+  )
+  internal fun refresh() {
+    registry.timer(refreshDurationId).record {
+      log.info("Refreshing plugin cache")
+
+      updateManager.refresh()
+
+      val newCache = updateManager.plugins
+        .map {
+          val plugin = DeckPluginVersion(it.id, selectPluginVersion(it))
+          val path = getOrDownload(plugin.id, plugin.version)
+          PluginCacheEntry(plugin, path)
+        }
+        .filter {
+          // TODO(rz): Once bundles support manifest artifacts, we can just inspect those
+          //  for a deck plugin rather than downloading everything and seeing if a
+          //  index.js file exists.
+          it.path.resolve("index.js").toFile().exists()
+        }
+
+      cache.removeIf { !newCache.contains(it) }
+      cache.addAll(newCache)
+
+      cache.forEach {
+        registry.counter(versionsId.withPluginTags(it.plugin.id, it.plugin.version)).increment()
+      }
+
+      cachePopulated = true
+      log.info("Cached ${cache.size} deck plugins")
+    }
+  }
+
+  fun isCachePopulated(): Boolean = cachePopulated
+
+  fun getCache(): Set<PluginCacheEntry> {
+    return cache.toSet()
+  }
+
+  /**
+   * Get a previously downloaded plugin path, or download the plugin and cache the artifacts for subsequent requests.
+   */
+  fun getOrDownload(pluginId: String, pluginVersion: String): Path {
+    val cachePath = CACHE_ROOT_PATH.resolve("$pluginId/$pluginVersion")
+    if (!cachePath.toFile().isDirectory) {
+      registry.timer(downloadDurationId.withPluginTags(pluginId, pluginVersion)).record {
+        log.info("Downloading plugin '$pluginId@$pluginVersion'")
```

Comment:
How expensive would calling `updateManager.refresh()` here be?  

After fetching the manifest of plugins, Deck could hit a different instance when requesting an asset.  Because of the polling interval, there will be cases where the gate instance  doesn't know about the version of the plugin being requested.